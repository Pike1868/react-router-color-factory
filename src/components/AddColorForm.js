import React from "react";

/**
 * AddColorForm Component:
 *
 * Displays form that allows users to create a new color, by providing a name and selecting a hex color, adds those values to the colors list on submission
 *
 * Props:
 *  - formData (object): An object containing color details. Expected format: {name: "colorName", hex: "#hexValue"}.
 *  - handleChange (function): Event handler function for input changes, updates UI.
 *  - handleSubmit (function): Event handler function for form submission, adds the new color to the colors list.
 *
 */

const AddColorForm = ({ formData, handleChange, handleSubmit }) => {
  return (
    <div style={{ textAlign: "center" }}>
      <form onSubmit={handleSubmit}>
        <label htmlFor="name">Choose a color name</label>
        <input
          required
          id="name"
          name="name"
          type="text"
          placeholder="Color Name"
          value={formData ? formData.name : ""}
          onChange={handleChange}
          style={{ margin: "20px" }}
        />
        <label htmlFor="hex">Choose a color hex</label>
        <input
          required
          id="hex"
          name="hex"
          type="color"
          value={formData ? formData.hex : "#000000"}
          onChange={handleChange}
          style={{ margin: "20px" }}
        />
        <button type="submit">Add new color</button>
      </form>
    </div>
  );
};

export default AddColorForm;
