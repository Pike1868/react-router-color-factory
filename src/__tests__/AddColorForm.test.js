import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import AddColorForm from "../components/AddColorForm";

//Smoke Test
test("AddColorForm renders without crashing", () => {
  render(
    <MemoryRouter>
      <AddColorForm />
    </MemoryRouter>
  );
});

//Snapshot Test
test("AddColorForm matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <AddColorForm />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

describe("AddColorForm Component Tests", () => {
  const handleChange = jest.fn();
  const handleSubmit = jest.fn();
  const initialFormData = { name: "", hex: "#000000" };

  test("renders with initial form data", () => {
    render(
      <AddColorForm
        formData={initialFormData}
        handleChange={handleChange}
        handleSubmit={handleSubmit}
      />
    );
    const nameInput = screen.getByRole("textbox");
    const hexInput = screen.getByLabelText("Choose a color hex");

    expect(nameInput.value).toBe(initialFormData.name);
    expect(hexInput.value).toBe(initialFormData.hex);
  });

  test("calls handleChange on input change", () => {
    render(
      <AddColorForm
        formData={initialFormData}
        handleChange={handleChange}
        handleSubmit={handleSubmit}
      />
    );
    const nameInput = screen.getByRole("textbox");
    fireEvent.change(nameInput, { target: { value: "purple" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  test("calls handleSubmit on form submission", () => {
    render(
      <AddColorForm
        formData={initialFormData}
        handleChange={handleChange}
        handleSubmit={handleSubmit}
      />
    );
    const submitButton = screen.getByRole("button", { name: "Add new color" });
    fireEvent.click(submitButton);
    expect(handleSubmit).toHaveBeenCalledTimes(1);
  });
});
