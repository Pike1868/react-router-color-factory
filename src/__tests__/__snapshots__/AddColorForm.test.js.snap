// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddColorForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="text-align: center;"
  >
    <form>
      <label
        for="name"
      >
        Choose a color name
      </label>
      <input
        id="name"
        name="name"
        placeholder="Color Name"
        required=""
        style="margin: 20px;"
        type="text"
        value=""
      />
      <label
        for="hex"
      >
        Choose a color hex
      </label>
      <input
        id="hex"
        name="hex"
        required=""
        style="margin: 20px;"
        type="color"
        value="#000000"
      />
      <button
        type="submit"
      >
        Add new color
      </button>
    </form>
  </div>
</DocumentFragment>
`;
